File: src/main/python/easy/No581_ShortestUnsortedContinuousSubarray.py
"""
581. 最短无序连续子数组
给定一个整数数组，你需要寻找一个连续的子数组，如果对这个子数组进行升序排序，那么整个数组都会变为升序排序。

你找到的子数组应是最短的，请输出它的长度。

示例 1:

输入: [2, 6, 4, 8, 10, 9, 15]
输出: 5
解释: 你只需要对 [6, 4, 8, 10, 9] 进行升序排序，那么整个表都会变为升序排序。
说明 :

输入的数组长度范围在 [1, 10,000]。
输入的数组可能包含重复元素 ，所以升序的意思是<=。

"""
from typing import List


class Solution:
    def findUnsortedSubarray(nums: List[int]) -> int:
        """
        先找到不递增的min；和递减得到的最小元素
        之后在不连续串内与这些边界值比较，寻找出更极端的边界值

        :param nums:
        :return:
        """
        n=len(nums)
        min = 2147483647
        max = -2147483648
        flag = False

        for i in range(1, n):
            if nums[i] < nums[i - 1]:
                flag = True
            if flag and min > nums[i]:
                min = nums[i]
        if not flag:
            return 0
        flag = False
        for i in range(n - 2, -1, -1):
            if nums[i] > nums[i + 1]:
                flag = True
            if flag and nums[i] > max:
                max = nums[i]

        _from = 0
        for i in range(0, n):
            if min < nums[i]:
                _from = i
                break

        _end = len(nums) - 1
        for i in range(len(nums) - 1, -1, -1):
            if max > nums[i]:
                _end = i
                break
        print(min, max)
        print(_from, _end)

        return 0 if _end - _from < 0 else _end - _from + 1

    if __name__ == '__main__':
        print(findUnsortedSubarray(nums=[1, 3, 2, 4, 5]))
        print(findUnsortedSubarray(nums=[1, 2, 3, 4]))

File: src/main/java/array/No581_ShortestUnsortedContinuousSubarray.java
package array;

/**
 * 581. 最短无序连续子数组
 * 给你一个整数数组 nums ，你需要找出一个 连续子数组 ，如果对这个子数组进行升序排序，那么整个数组都会变为升序排序。
 * 请你找出符合题意的 最短 子数组，并输出它的长度。
 * 示例 1：
 * 输入：nums = [2,6,4,8,10,9,15]
 * 输出：5
 * 解释：你只需要对 [6, 4, 8, 10, 9] 进行升序排序，那么整个表都会变为升序排序。
 * 示例 2：
 * 输入：nums = [1,2,3,4]
 * 输出：0
 * 示例 3：
 * 输入：nums = [1]
 * 输出：0
 * 提示：
 * 1 <= nums.length <= 104
 * -105 <= nums[i] <= 105
 */
public class No581_ShortestUnsortedContinuousSubarray {

    /**
     * 优先掌握
     * 这个算法背后的思想是无序子数组中最小元素的正确位置可以决定左边界，最大元素的正确位置可以决定右边界。
     * @param nums
     * @return
     */
    public int findUnsortedSubarray(int[] nums) {
        int len = nums.length;
        int left = 0, right = len - 1;
        while (left < right && nums[left] <= nums[left + 1]) left++;
        while (left < right && nums[right] >= nums[right - 1]) right--;
        int min = nums[left], max = nums[right];
        int l = left, r = right;
        for (int k = l; k <= r; k++) {
            if (nums[k] < min) {
                while (left >= 0 && nums[left] > nums[k]) left--;
                min = left >= 0 ? nums[left] : Integer.MIN_VALUE;
            }
            if (nums[k] > max) {
                while (right < len && nums[right] < nums[k]) right++;
                max = right < len ? nums[right] : Integer.MAX_VALUE;
            }
        }
        return right == left ? 0 : (right - 1) - (left + 1) + 1;
    }


    public int findUnsortedSubarray2(int[] nums) {
        int len = nums.length;
        int left = -1, min = Integer.MAX_VALUE;
        int right = -1, max = Integer.MIN_VALUE;
        for (int i = 0; i < len; i++) {
            if (nums[i] < max) right = i;
            else max = nums[i];

            if (nums[len - 1 - i] > min) left = len - 1 - i;
            else min = nums[i];
        }
        return right == -1 ? 0 : right - left + 1;
    }

    int MIN = -100005, MAX = 100005;



}
